File: app/src/main/java/org/michaelbel/moviemade/ui/activity/MovieActivity.java
package org.michaelbel.moviemade.ui.activity;

import android.content.Intent;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;
import android.widget.TextView;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;

import org.michaelbel.moviemade.R;
import org.michaelbel.moviemade.Url;
import org.michaelbel.moviemade.browser.Browser;
import org.michaelbel.moviemade.mvp.base.BaseActivity;
import org.michaelbel.moviemade.ui.fragment.MovieFragment;
import org.michaelbel.moviemade.ui.view.BackdropView;
import org.michaelbel.moviemade.ui.view.appbar.AppBarState;
import org.michaelbel.moviemade.ui.view.appbar.AppBarStateChangeListener;
import org.michaelbel.tmdb.v3.json.Movie;

import java.util.Locale;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import butterknife.BindView;
import butterknife.ButterKnife;

@SuppressWarnings("all")
public class MovieActivity extends BaseActivity {

    public Movie movie;

    private Menu actionMenu;
    private MenuItem menu_share;
    private MenuItem menu_tmdb;
    private MenuItem menu_imdb;

    @BindView(R.id.toolbar)
    public Toolbar toolbar;

    @BindView(R.id.app_bar)
    public AppBarLayout appBar;

    @BindView(R.id.toolbar_title)
    public TextView toolbarTitle;

    @BindView(R.id.backdrop_image)
    public BackdropView backdropImage;

    @BindView(R.id.collapsing_layout)
    public CollapsingToolbarLayout collapsingToolbarLayout;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie);
        ButterKnife.bind(this);

        movie = new Movie();
        movie.id =  getIntent().getIntExtra("id", 0);
        movie.title = getIntent().getStringExtra("title");
        movie.backdropPath = getIntent().getStringExtra("backdropPath");
        movie.posterPath = getIntent().getStringExtra("posterPath");
        movie.overview = getIntent().getStringExtra("overview");
        movie.voteAverage = getIntent().getFloatExtra("voteAverage", 0);
        movie.voteCount = getIntent().getIntExtra("voteCount", 0);
        movie.adult = getIntent().getBooleanExtra("adult", false);
        movie.video = getIntent().getBooleanExtra("video", false);
        movie.releaseDate = getIntent().getStringExtra("releaseDate");
        //movie.genreIds = getIntent().getIntegerArrayListExtra("genreIds");
        movie.originalTitle = getIntent().getStringExtra("originalTitle");
        movie.originalLanguage = getIntent().getStringExtra("originalLanguage");
        movie.popularity = getIntent().getDoubleExtra("popularity", 0);

        MovieFragment fragment = (MovieFragment) getSupportFragmentManager().findFragmentById(R.id.fragment);
        fragment.presenter.setMovieDetailsFromExtra(movie);
        fragment.presenter.loadMovieDetails(movie.id);

        getWindow().setStatusBarColor(ContextCompat.getColor(this, R.color.primary));
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        getWindow().setStatusBarColor(ContextCompat.getColor(this, R.color.transparent20));

        toolbar.setNavigationIcon(R.drawable.ic_arrow_back);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(view -> finish());
        toolbar.setTitle(null);

        appBar.addOnOffsetChangedListener(new AppBarStateChangeListener() {
            @Override
            public void onStateChanged(AppBarLayout appBarLayout, AppBarState state) {
                if (state == AppBarState.COLLAPSED) {
                    toolbarTitle.setText(movie.title);
                } else {
                    toolbarTitle.setText(null);
                }
            }
        });

        toolbarTitle.setText(movie.title);
        backdropImage.setImage(movie.backdropPath);

        collapsingToolbarLayout.setContentScrimColor(ContextCompat.getColor(this, R.color.primary));
        collapsingToolbarLayout.setStatusBarScrimColor(ContextCompat.getColor(this, android.R.color.transparent));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        actionMenu = menu;
        menu_share = menu.add(R.string.Share).setIcon(R.drawable.ic_anim_share).setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_IF_ROOM);
        menu_tmdb = menu.add(R.string.ViewOnTMDb).setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_NEVER);
        //menu_imdb = menu.add(R.string.ViewOnIMDb).setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_NEVER);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item == menu_share) {
            Drawable icon = actionMenu.getItem(0).getIcon();
            if (icon instanceof Animatable) {
                ((Animatable) icon).start();
            }

            Intent intent = new Intent(Intent.ACTION_SEND);
            intent.setType("text/plain");
            intent.putExtra(Intent.EXTRA_TEXT, String.format(Locale.US, Url.TMDB_MOVIE, movie.id));
            startActivity(Intent.createChooser(intent, getString(R.string.ShareVia)));
        } else if (item == menu_tmdb) {
            Browser.openUrl(this, String.format(Locale.US, Url.TMDB_MOVIE, movie.id));
        } /*else if (item == menu_imdb) {
            Browser.openUrl(context, String.format(Locale.US, Url.IMDB_MOVIE, movie.imdbId));
        }*/

        return true;
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/features/movie/MovieActivity.kt
package org.michaelbel.moviemade.presentation.features.movie

import android.content.SharedPreferences
import android.os.Bundle
import android.view.View.*
import android.view.WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS
import androidx.core.content.ContextCompat
import androidx.fragment.app.commitNow
import by.kirich1409.viewbindingdelegate.viewBinding
import com.google.android.gms.ads.MobileAds
import com.google.android.material.appbar.AppBarLayout
import com.squareup.picasso.Picasso
import dagger.hilt.android.AndroidEntryPoint
import org.michaelbel.data.remote.model.Movie
import org.michaelbel.moviemade.R
import org.michaelbel.moviemade.core.TmdbConfig
import org.michaelbel.moviemade.core.local.SharedPrefs.KEY_ACCOUNT_BACKDROP
import org.michaelbel.moviemade.core.local.SharedPrefs.KEY_SESSION_ID
import org.michaelbel.moviemade.databinding.ActivityMovieBinding
import org.michaelbel.moviemade.ktx.displayWidth
import org.michaelbel.moviemade.ktx.toDp
import org.michaelbel.moviemade.ktx.toast
import org.michaelbel.moviemade.presentation.ContainerActivity.Companion.EXTRA_MOVIE
import org.michaelbel.moviemade.presentation.common.appbar.AppBarStateChangeListener
import org.michaelbel.moviemade.presentation.common.base.BaseActivity
import org.michaelbel.moviemade.presentation.listitem.TextListItem
import org.michaelbel.moviemade.presentation.widget.BottomSheetDialog
import javax.inject.Inject

@AndroidEntryPoint
class MovieActivity: BaseActivity(R.layout.activity_movie) {

    val binding: ActivityMovieBinding by viewBinding(R.id.parentLayout)

    private lateinit var movie: Movie
    private lateinit var fragment: MovieFragment

    @Inject lateinit var preferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        MobileAds.initialize(this)

        movie = intent.getSerializableExtra(EXTRA_MOVIE) as Movie

        fragment = MovieFragment.newInstance(movie)
        supportFragmentManager.commitNow { replace(binding.container.id, fragment) }

        window.addFlags(FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = ContextCompat.getColor(this, R.color.transparent40)

        binding.toolbar.setNavigationIcon(R.drawable.ic_arrow_back)
        setSupportActionBar(binding.toolbar)
        binding.toolbar.setNavigationOnClickListener {finish()}
        binding.toolbar.title = null

        binding.appBar.addOnOffsetChangedListener(object: AppBarStateChangeListener() {
            override fun onStateChanged(appBarLayout: AppBarLayout, state: AppBarState) {
                binding.toolbarTitle.text = if (state === AppBarState.COLLAPSED) movie.title else null
                binding.toolbarTitle.isSelected = state === AppBarState.COLLAPSED
            }
        })

        /**
         * todo
         * supportActionBar?.title пока не получается использовать,
         * так как его перехватывает collapsingToolbarLayout.title.
         */
        binding.toolbarTitle.text = movie.title

        if (!movie.backdropPath.isNullOrEmpty()) {
            Picasso.get().load(TmdbConfig.image(movie.backdropPath as String)).resize(displayWidth, 220F.toDp(applicationContext)).into(binding.cover)
        }

        binding.collapsingLayout.setContentScrimColor(ContextCompat.getColor(this, R.color.primary))

        binding.cover.setOnLongClickListener {
            val sessionId: String? = preferences.getString(KEY_SESSION_ID, "")

            if (!sessionId.isNullOrEmpty() && !movie.backdropPath.isNullOrEmpty()) {
                val dialog = BottomSheetDialog(top = 8F, bottom = 8F)

                val listItem = TextListItem(TextListItem.Data(text = R.string.set_as_background, divider = false, medium = false))
                listItem.listener = object: TextListItem.Listener {
                    override fun onClick() {
                        preferences.edit().putString(KEY_ACCOUNT_BACKDROP, TmdbConfig.image(movie.backdropPath as String)).apply()
                        toast(R.string.msg_done)
                        dialog.dismiss()
                    }
                }

                dialog.addItem(listItem)
                dialog.show(supportFragmentManager, dialog.tag + "1")
                return@setOnLongClickListener true
            }
            return@setOnLongClickListener false
        }
    }
}